### PR TITLE
Audio and Image comments

### PR DESCRIPTION
**Pull request by MattVZDeakin**

### CODE REVIEW

File: src/app/common/services/task-service.coffee [hunk 1]
```diff
@@ -526,5 +528,41 @@ angular.module("doubtfire.common.services.tasks", [])
...
+
+    form = new FormData()
+    form.append 'type', commentType
+    form.append 'project_id', task.project().project_id
+    form.append 'task_definition_id', task.task_definition_id
+
+    if commentType == "image"
+      form.append 'attachment', mediaURL[0]
+    else if commentType == "audio"
+      xhr = new XMLHttpRequest()
```
Comment (macite) [inline src/app/common/services/task-service.coffee]: Why are we using an XHR here and the resource below? Do we need both?

**Comment by macite** [comment]
@alexcu and @jakerenzella if you get a chance can you look over this as well?

File: src/app/common/markdown-editor/markdown-editor.coffee [hunk 2]
```diff
@@ -31,3 +55,273 @@ angular.module('doubtfire.common.markdown-editor', [])
...
+              taskService.addMediaComment CommentResourceService.task, mediaUrl, $scope.currentCommentType
+              return
+            return
+          return
+      #---------------------------------------------
+      # pass initRecording a cb to call when stream is ready
+      initRecording = ->
+
+        if navigator.mediaDevices and navigator.mediaDevices.getUserMedia
+          console.log 'getUserMedia supported.'
```
Comment (jakerenzella) [inline src/app/common/markdown-editor/markdown-editor.coffee]: This line should be removed

File: src/app/common/audio-player/audio-player.coffee [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  controller: ($scope, CommentResourceService, Task, $sce, $http, $q) ->
+    DEFAULT_HEIGHT = 300
+
+    $scope.loadPlayer = false
+
+    $scope.commentAudioUrl = CommentResourceService.commentAudioUrl
+
+    #=========================================================================
+    getAudioComment = (comment, audioElement) ->
+      audioUrl = $sce.trustAsResourceUrl(Task.generateCommentsAttachmentUrl($scope.project, $scope.task, comment))
```
[the remark this answers is not in the document]
Comment (MattVZDeakin) [reply src/app/common/audio-player/audio-player.coffee]: Not sure how to go about doing that.

File: src/app/common/modals/comments-modal/comments-modal.coffee [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+.factory("CommentsModal", ($modal) ->
+  CommentsModal = {}
+  CommentsModal.show = ->
+    $modal.open
+      templateUrl: 'common/modals/comments-modal/comments-modal.tpl.html'
+      controller: 'CommentsModalCtrl'
+      size: 'lg'
+  CommentsModal
+)
+.controller("CommentsModalCtrl", ($scope, $modalInstance, CommentResourceService) ->
```
Comment (macite) [inline src/app/common/modals/comments-modal/comments-modal.coffee]: Pass comment or image...

Comment (macite) [reply src/app/common/modals/comments-modal/comments-modal.coffee]: Remove CommentResourceService - as just used to add global variables

File: src/app/common/audio-player/audio-player.coffee [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  restrict: 'E'
+  replace: true
+  templateUrl: 'common/audio-player/audio-player.tpl.html'
+
+  controller: ($scope, CommentResourceService, Task, $sce, $http, $q) ->
+    DEFAULT_HEIGHT = 300
+
+    $scope.loadPlayer = false
+
+    $scope.commentAudioUrl = CommentResourceService.commentAudioUrl
```
Comment (macite) [inline src/app/common/audio-player/audio-player.coffee]: Pass in audio url - so we can remove CommentResourceService (globals)

File: src/app/common/audio-player/audio-player.coffee [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+      isPlaying = false
+      btnPlay = document.getElementById('btnPlay-' + index)
+      audioElement = document.getElementById('audio-element-' + index)
+      getAudioComment(comment,audioElement)
+      audioElement.crossOrigin = "anonymous"
+      canvasElement = document.getElementById('audio-visualiser-' + index)
+
+      CommentResourceService.setupAudioContext()
+      audioCtx = CommentResourceService.audioContext
```
Comment (macite) [inline src/app/common/audio-player/audio-player.coffee]: Keep this aspect of CommentResourceService - keeping track of the context and testing for capability/support

File: src/app/common/markdown-editor/markdown-editor.coffee [hunk 7]
```diff
@@ -11,8 +11,32 @@ angular.module('doubtfire.common.markdown-editor', [])
...
+      image:
+        extensions: ['png', 'bmp', 'jpeg', 'jpg']
+        name:       'image'
+        icon:       'fa-file-image-o'
+
+    $scope.commentType = {TEXT : "text", AUDIO : "audio", IMAGE : "image"} # Enum for keeping track of comment type
+    $scope.currentCommentType = $scope.commentType.TEXT # Current value of comment type
+
+    $scope.$watch 'currentCommentType', ->
+      CommentResourceService.setCommentType($scope.currentCommentType)
```
Comment (macite) [inline src/app/common/markdown-editor/markdown-editor.coffee]: Remove this...

File: src/app/common/markdown-editor/markdown-editor.coffee [hunk 8]
```diff
@@ -31,3 +55,267 @@ angular.module('doubtfire.common.markdown-editor', [])
...
+
+    #============================================================================
+    $scope.clearEnqueuedUpload = (upload) ->
+      upload.model = null
+      refreshShownUploadZones()
+
+    #============================================================================
+    # Upload image files as comments to a given task
+    $scope.postImageComment = ->
+      taskService.addMediaComment(CommentResourceService.task, $scope.upload.model, $scope.currentCommentType)
```
Comment (macite) [inline src/app/common/markdown-editor/markdown-editor.coffee]: Remove CommentResourceService - we are on the task already

File: src/app/common/markdown-editor/markdown-editor.coffee [hunk 9]
```diff
@@ -31,3 +55,267 @@ angular.module('doubtfire.common.markdown-editor', [])
...
+      
+      btnRecord = document.getElementById('btnRecord')
+      btnPlay = document.getElementById('btnPlay')
+      btnSend = document.getElementById('btnSend')
+      btnDelete = document.getElementById('btnDelete')
+      canvas = document.getElementById('audio-visualiser')
+      audio = document.getElementById('audioPlayer')
+
+      CommentResourceService.setupAudioContext()
+      audioCtx = CommentResourceService.audioContext
```
Comment (macite) [inline src/app/common/markdown-editor/markdown-editor.coffee]: Keep - organising context

File: src/app/common/markdown-editor/markdown-editor.coffee [hunk 10]
```diff
@@ -31,3 +55,267 @@ angular.module('doubtfire.common.markdown-editor', [])
...
+            type = blob.type
+            fileReader = new FileReader
+            fileReader.readAsDataURL blob
+            fileReader.addEventListener 'loadend', ->
+              fileData = fileReader.result
+              downloadMediaContent = new Blob([ fileData ], 'type': type)
+              downloadLink = document.getElementById('btnSend')
+              mediaUrl = window.URL.createObjectURL(downloadMediaContent)
+              downloadLink.href = mediaUrl
+              taskService.addMediaComment CommentResourceService.task, downloadMediaContent, $scope.currentCommentType
```
Comment (macite) [inline src/app/common/markdown-editor/markdown-editor.coffee]: Should already have task here - remove CommentResourceService

File: src/app/common/modals/comments-modal/comments-modal.coffee [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+  $scope.commentImageUrl = CommentResourceService.commentImageUrl
```
Comment (macite) [inline src/app/common/modals/comments-modal/comments-modal.coffee]: Also refactor here...

File: src/app/common/services/comment-service.coffee [hunk 12]
```diff
@@ -0,0 +1,37 @@
+angular.module("doubtfire.common.services.comments", [])
+#
+# Services for transferring comment resources
+#
+.factory("CommentResourceService", ->
```
Comment (macite) [inline src/app/common/services/comment-service.coffee]: Refactor - remove globals used to pass task and comment related urls around.

Add comment related URLs into the Comment itself
Tasks should be available
Pass via scope if needed

Keep the support for context and checking capability/support from browser

File: src/app/tasks/task-comments-viewer/task-comments-viewer.coffee [hunk 13]
```diff
@@ -29,34 +30,55 @@ angular.module("doubtfire.tasks.task-comments-viewer", [])
         project_id: $scope.project.project_id,
         task_definition_id: $scope.task.task_definition_id
       }, (response) ->
-        $scope.task.comments = _.map(response, taskService.mapComment)
+        $scope.task.comments = _.map(response, taskService.mapComment) #in the HTML, the mapped task.comments are displayed
         $scope.task.num_new_comments = 0
         scrollDown()
         $scope.focus?() if $scope.refocusOnTaskChange
 
+        CommentResourceService.setTask($scope.task)
```
Comment (macite) [inline src/app/tasks/task-comments-viewer/task-comments-viewer.coffee]: Remove global tracking... CommentResourceService

File: src/app/tasks/task-comments-viewer/task-comments-viewer.coffee [hunk 14]
```diff
@@ -29,34 +30,55 @@ angular.module("doubtfire.tasks.task-comments-viewer", [])
...
     # Automatically scroll the inner div to the bottom of comments
     scrollDown = ->
       $timeout ->
         objDiv = document.querySelector("task-comments-viewer .panel-body")
         wrappedResult = angular.element(objDiv)
         wrappedResult[0].scrollTop = wrappedResult[0].scrollHeight
 
+    $scope.openCommentsModal = (comment)->
+      imageUrl = $sce.trustAsResourceUrl(Task.generateCommentsAttachmentUrl($scope.project, $scope.task, comment))
+      CommentResourceService.setImageUrl(imageUrl)
```
Comment (macite) [inline src/app/tasks/task-comments-viewer/task-comments-viewer.coffee]: Also remove here...

File: src/app/tasks/task-comments-viewer/task-comments-viewer.coffee [hunk 15]
```diff
@@ -29,34 +30,55 @@ angular.module("doubtfire.tasks.task-comments-viewer", [])
...
     $scope.addComment = ->
       $scope.comment.text = $scope.comment.text.trim()
-      taskService.addComment $scope.task, $scope.comment.text,
-        (success) ->
-          $scope.comment.text = ""
-          analyticsService.event "View Task Comments", "Added new comment"
-          scrollDown()
-        (response) ->
-          alertService.add("danger", response.data.error, 2000)
+      taskService.addComment $scope.task, $scope.comment.text, CommentResourceService.commentType,
```
Comment (macite) [inline src/app/tasks/task-comments-viewer/task-comments-viewer.coffee]: Move selection of text/image/audio/video here... then we know what we need to add. Remove from CommentResourceService

Markdown editor needs to just handle markdown

**Pull request merged**